osu.Game.Rulesets.Osu/Difficulty/Skills/Speed.cs

```diff
@@ -20,51 +21,168 @@ public class Speed : OsuStrainSkill
+
+                    if (effectiveRatio > 0.5)
+                        effectiveRatio = 0.5 + (effectiveRatio - 0.5) * 5; // large buff for 1/3 -> 1/4 type transitions.
+
+                    effectiveRatio *= currHistoricalDecay; // scale with time
+
+                    if (firstDeltaSwitch)
+                    {
+                        if (Precision.AlmostEquals(prevDelta, currDelta, 15))
+                        {
+                            islandSize++; // island is still progressing, count size.
+                        }
+                        else
+                        {
+                            if (islandSize > 6)
+                                islandSize = 6;
+
+                            if (Previous[i - 1].BaseObject is Slider) // bpm change is into slider, this is easy acc window
+                                effectiveRatio *= 0.25;
+
+                            if (Previous[i].BaseObject is Slider) // bpm change was from a slider, this is easier typically than circle -> circle
+                                effectiveRatio *= 0.5;
+
+                            if (previousIslandSize == islandSize) // repeated island size (ex: triplet -> triplet)
+                                effectiveRatio *= 0.25;
+
+                            if (prevPrevDelta > prevDelta + 10 && prevDelta > currDelta + 10) // previous increase happened a note ago, 1/1->1/2-1/4, dont want to buff this.
+                                effectiveRatio *= 0.125;
+
+                            rhythmComplexitySum += effectiveRatio;
+
+                            previousIslandSize = islandSize; // log the last island size.
+
+                            if (prevDelta * 1.25 < currDelta) // we're slowing down, stop counting
+                                firstDeltaSwitch = false; // if we're speeding up, this stays true and  we keep counting island size.
+
+                            islandSize = 0;
+                        }
+                    }
+                    else if (prevDelta > 1.25 * currDelta) // we want to be speeding up.
+                    {
+                        // Begin counting island until we change speed again.
+                        firstDeltaSwitch = true;
+                        islandSize = 0;
+                    }
+                }
+            }
+
+            return Math.Sqrt(4 + rhythmComplexitySum * rhythm_multiplier) / 2; //produces multiplier that can be applied to strain. range [1, infinity)
+        }
+
+        private double tapStrainOf(DifficultyHitObject current, double speedBonus)
+        {
+            if (current.BaseObject is Spinner)
+                return 0;
+
+            var osuCurrObj = (OsuDifficultyHitObject)current;
+
+            return speedBonus / osuCurrObj.StrainTime;
+        }
+
+        private double movementStrainOf(DifficultyHitObject current, double speedBonus)
```

Comment:
At this point I feel like this shouldn't be here at all. Maybe we can move this into aim or have tap as a separate skill? I'm not talking about reworking skills as you love doing, I'm talking about just splitting things up.

Reply:
This is a massively large change that is out of the scope of this rework. It can be done at a later time.